lex_test.go

```diff
@@ -0,0 +1,332 @@
+package mf2
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_lex(t *testing.T) {
+	t.Parallel()
+
+	var (
+		tokenEOF             = mk(itemEOF, "")
+		tokenExpressionOpen  = mk(itemExpressionOpen, "{")
+		tokenExpressionClose = mk(itemExpressionClose, "}")
+	)
+
+	for _, test := range []struct {
+		name     string
+		input    string // MessageFormat2 formatted string
+		expected []item
+	}{
+		{
+			name:     "empty simple message",
+			input:    "",
+			expected: []item{tokenEOF},
+		},
+		{
+			name:  "text",
+			input: `escaped text: \\ \} \{`,
+			expected: []item{
+				mk(itemText, `escaped text: \ } {`),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "function",
+			input: "{:rand}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemFunction, "rand"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+
+		{
+			name:  "opening function",
+			input: "{+button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemOpeningFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "closing function",
+			input: "{-button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemClosingFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "opening and closing functions",
+			input: "{+button}Submit{-button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemOpeningFunction, "button"),
+				tokenExpressionClose,
+				mk(itemText, "Submit"),
+				tokenExpressionOpen,
+				mk(itemClosingFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "variable",
+			input: "{$count :math:round}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemVariable, "count"),
+				mk(itemWhitespace, " "),
+				mk(itemFunction, "math:round"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "text with variable",
+			input: "Hello, {$guest}!",
+			expected: []item{
+				mk(itemText, "Hello, "),
+				tokenExpressionOpen,
+				mk(itemVariable, "guest"),
+				tokenExpressionClose,
+				mk(itemText, "!"),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "empty quoted literal",
+			input: "{||}",
+			expected: []item{
+				mk(itemExpressionOpen, "{"),
+				mk(itemLiteral, ""),
+				mk(itemExpressionClose, "}"),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "quoted literal",
+			input: "{|Hello, world!| :uppercase}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemLiteral, "Hello, world!"),
```

Comment:
Not sure if differentiation is required. What are the benefits?

Reply:
Not losing information when marshaling from MF2 to AST and back.

Reply:
The only lost information is about how it was represented in the original syntax.

When encoding back to MF2, you still have to decide to use unquoted or quoted - `Hello` or `|Hello|`.

E.g. you have unquoted `Hello` in the original. Later, it was changed to `Hello, world!` in the AST. You cannot use unquoted anymore. You have to use quoted literal now - `|Hello, world!|`.

Reply:
It would be nice to keep the original syntax intact for testing when no modifications are made.

Reply:
What scenarious are you talking about, @klavinsjanis?

It would make it harder for AST consumer. It enforces decision making onto the consumer when making changes to AST - to use quoted or unquoted.

Reply:
ok, that makes sense. I was just wondering if we will get the same result after unmarshal & marshal back to MF2.
MF2 "{hello :uppercase}" -> AST ->  MF2  "{hello :uppercase}"

Reply:
For the sake of simplicity, we might use quoted literals when marshalling. We'll see.